src/js/components/ServiceForm.js

```diff
@@ -0,0 +1,230 @@
+      return;
+    }
+
+    if (isBlur) {
+      this.validateForm();
+      this.props.onChange(this.getModel());
+      return;
+    }
+
+    // The cleartimeout is there to debounce the validation. And to make
+    // sure it is only run once.
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    this.timer = setTimeout(() => {
+      this.validateForm();
+      this.props.onChange(this.getModel());
+    });
+  }
+
+  handleExternalSubmit() {
+    this.validateForm();
+    return this.getModel();
+  }
+
+  handleRemoveRow(definition, prop, id) {
+    FormUtil.removePropID(definition, prop, id);
+    this.forceUpdate();
+  }
+
+  handleAddRow(prop, definition, newDefinition) {
+    let propID = Util.uniqueID(prop);
+    newDefinition = FormUtil.getMultipleFieldDefinition(
+      prop,
+      propID,
+      newDefinition
+    );
+    newDefinition.push(
+      this.getRemoveVariableButton(definition, prop, propID)
+    );
+    definition.splice(definition.length - 1, 0, newDefinition);
+    this.forceUpdate();
+  }
+
+  transformPorts(definition, startValues) {
+    let generalDefinition = definition.environmentVariables.definition;
+    let instancesIndex = 0;
+    let instancesDefinition = generalDefinition.find(function (fieldDefinition, index) {
+      if (fieldDefinition.name === 'ports') {
+        instancesIndex = index;
+        return true;
+      }
+
+      return false;
+    });
+
+    let transformOptions = {
+      prop: 'ports',
+      parentDefinition: generalDefinition,
+      itemDefinition: instancesDefinition.itemShape.definition,
+      index: instancesIndex,
+      getRemoveButton: this.getRemoveVariableButton,
+      getNewRowButton: this.getAddNewRowButton,
+      startValues
+    };
+
+    FormUtil.transformDefinition(transformOptions);
+  }
+
+  getAddNewRowButton(prop, generalDefinition, definition) {
+    return (
+      <button
+        className="button"
+        onClick={
+          this.handleAddRow.bind(this, prop, generalDefinition, definition)
+        }>
+        Add New Variable
+      </button>
+    );
+  }
+
+  getRemoveVariableButton(generalDefinition, prop, id) {
+    return (
+      <button
+        className="button"
+        key={prop + id}
+        onClick={this.handleRemoveRow.bind(this, generalDefinition, prop, id)}>
+        X
+      </button>
+    );
+  }
+
+  getModel() {
+    let model = this.triggerTabFormSubmit();
+    model = SchemaFormUtil.processFormModel(model, this.multipleDefinition);
+    model.environmentVariables = FormUtil.modelToCombinedProps(
+      'ports',
+      model.environmentVariables
+    );
+
+    return model;
+  }
+
+  getNewDefinition() {
+    let {model, schema} = this.props;
+    let definition = SchemaUtil.schemaToMultipleDefinition(
+      schema, this.getSubHeader, this.getLabel
+    );
+
+    if (model) {
+      SchemaFormUtil.mergeModelIntoDefinition(model, definition);
+    }
+
+    this.transformPorts(definition, model.environmentVariables.ports);
```

Comment:
Could this be automatic? I mean for all array typed fields, this would reduce redundancy when we add new array typed fields. That we have to add it in multiple places.